plugin/pkg/admission/certificates/admission.go

```diff
@@ -0,0 +1,131 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"k8s.io/klog"
+
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninit "k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+
+	api "k8s.io/kubernetes/pkg/apis/certificates"
+)
+
+// PluginName is a string with the name of the plugin
+const PluginName = "CertificateApproval"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return newPlugin(), nil
+	})
+}
+
+// Plugin holds state for and implements the admission plugin.
+type Plugin struct {
+	*admission.Handler
+	authz authorizer.Authorizer
+}
+
+// SetAuthorizer sets the authorizer.
+func (p *Plugin) SetAuthorizer(authz authorizer.Authorizer) {
+	p.authz = authz
+}
+
+// ValidateInitialization ensures an authorizer is set.
+func (p *Plugin) ValidateInitialization() error {
+	if p.authz == nil {
+		return fmt.Errorf("%s requires an authorizer", PluginName)
+	}
+	return nil
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+var _ genericadmissioninit.WantsAuthorizer = &Plugin{}
+
+// newPlugin creates a new CSR approval admission plugin
+func newPlugin() *Plugin {
+	return &Plugin{
+		Handler: admission.NewHandler(admission.Update),
+	}
+}
+
+// Validate verifies that the requesting user has permission to approve
+// CertificateSigningRequests for the specified signerName.
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Ignore all calls to anything other than 'certificatesigningrequests/approval'.
+	// Ignore all operations other than UPDATE.
+	if a.GetSubresource() != "approval" ||
+		a.GetResource().GroupResource() != api.Resource("certificatesigningrequests") ||
+		a.GetOperation() != admission.Update {
+		return nil
+	}
+
+	csr := a.GetObject().(*api.CertificateSigningRequest)
+
+	// if signerName is not set, the approving user must have permission to
+	// approve *all* signerNames (i.e. no 'resourceNames' attribute provided).
+	// TODO: should we apply defaulting logic here and test for legacy-unknown,
+	//  or alternatively always allow requests if the signerName is empty as it
+	//  implies we are speaking to a non-signerName aware client
+	signerName := ""
+	if csr.Spec.SignerName != nil {
+		signerName = *csr.Spec.SignerName
+	}
```

Comment:
Does regular defaulting occur when a subresource endpoint is hit? i.e. is it safe for us to assume that the defaulting that's applied to CSRs will have been applied by this point even if the user has hit the `certificatesigningrequests/approval` subresource? If yes, then we can remove this mitigation altogether and directly dereference `csr.Spec.SignerName` as we can be certain that the defaulting logic has set this field to *some value*.

Reply:
This should be a non-issue since defaulting occurs on create.